packages/frontend/src/redux/reducers/sign/index.js

```diff
@@ -25,43 +30,62 @@ const sign = handleActions({
-                gasLimit: allActions
-                    .filter(a => Object.keys(a)[0] === 'functionCall')
-                    .map(a => a.functionCall.gas)
-                    .reduce((totalGas, gas) => totalGas.add(gas), new BN(0)).toString(),
+                gasLimit: calculateGasLimit(allActions),
                 gasPrice: '' // TODO: Where to get gas price?
             },
             sensitiveActionsCounter: allActions
                 .filter(a => ['deployContract', 'stake', 'deleteAccount'].indexOf(Object.keys(a)[0]) > -1)
                 .length
         };
     },
-    [signAndSendTransactions]: (state, { error, payload, ready }) => {
-
-        if (!ready) {
-            return {
-                ...state,
-                status: 'needs-confirmation'
-            };
-        }
+    [handleSignTransactions.pending]: (state) => {
+        const { retryTxDirection, status } = state;
 
-        if (error) {
-            return {
-                ...state,
-                status: 'error',
-                error: payload
-            };
-        }
+        const transactions = status === SIGN_STATUS.RETRY_TRANSACTION
+            ? increaseGasForTransactions({ 
+                transactions: cloneDeep(state.transactions),
+                retryTxDirection: state.retryTxDirection || retryTxDirection }
+            )
+            : state.transactions;
 
         return {
             ...state,
-            status: 'success'
+            status: SIGN_STATUS.IN_PROGRESS,
+            transactions,
+            fees: {
+                ...state.transactions.fees,
+                gasLimit: calculateGasLimit(transactions.flatMap(t => t.actions))
+            }
         };
     },
-    [setSignTransactionStatus]: (state, { payload }) => {
+    [handleSignTransactions.fulfilled]: (state, { payload }) => ({
+        ...state,
+        status: SIGN_STATUS.SUCCESS,
+        transactionHashes: payload,
+        error: undefined
+    }),
+    [handleSignTransactions.rejected]: (state, { error }) => {
+        const retryTxDirection = error.message.includes('Exceeded the prepaid gas')
+            ? RETRY_TX.INCREASE
+            : undefined;
+        
+        const tryRetryTx = retryTxDirection && !state.transactions.every((t) => 
+            t.actions && t.actions.every((a) => a.functionCall && a.functionCall.gas && (
+                state.retryTxDirection === RETRY_TX.INCREASE 
+                && (
+                    a.functionCall.gas.gt(new BN(RETRY_TX.GAS.MAX))
+                    || a.functionCall.gas.eq(new BN(RETRY_TX.GAS.MAX))
+                )
+            ))
+        );
+        
         return {
             ...state,
-            status: payload.status
+            status: tryRetryTx
+                ? SIGN_STATUS.RETRY_TRANSACTION
+                : SIGN_STATUS.ERROR,
+            error,
+            retryTxDirection: state.retryTxDirection || retryTxDirection
```

Comment:
We should be able to just use `state.status` to know what to do now that there is only one direction to go. (🔪 `retryTxDirection`)

Reply:
Done.